src/hooks/transactions/useTransactionPreparation.ts

```diff
@@ -33,89 +27,21 @@ interface Props {
   isValidCall?: boolean;
 }
 
-interface FeeAndPayload {
-  payload: any;
-}
-
-export default function useTransactionPreparation({
-  input,
-  type,
-  weightInput,
-  isValidCall = true
-}: Props): FeeAndPayload {
-  const { areApiReady } = useLoadingApi();
-  const laneId = useLaneId();
-  const {
-    sourceChainDetails: {
-      apiConnection: { api: sourceApi }
-    },
-    targetChainDetails: { chain: targetChain }
-  } = useSourceTarget();
-  const { account } = useAccountContext();
-
-  const [payload, setPayload] = useState<null | {}>(null);
+export default function useTransactionPreparation({ input, type, weightInput, isValidCall = true }: Props) {
+  const { payload } = useTransactionContext();
   const { call, weight } = useTransactionType({ input, type, weightInput });
 
-  const { dispatchTransaction } = useUpdateTransactionContext();
-  const { estimatedFeeMethodName } = getSubstrateDynamicNames(targetChain);
+  const calculateFee = useEstimateFee();
+  const updatePayload = usePayload({ call, weight });
 
   useEffect(() => {
-    const calculateFee = async () => {
-      // Ignoring custom types missed for TS for now.
-      // Need to apply: https://polkadot.js.org/docs/api/start/typescript.user
-      // @ts-ignore
-      const payloadType = sourceApi.registry.createType('OutboundPayload', payload);
-      // @ts-ignore
-      const messageFeeType = sourceApi.registry.createType('MessageFeeData', {
-        lane_id: laneId,
-        payload: payloadType.toHex()
-      });
-
-      const estimatedFeeCall = await sourceApi.rpc.state.call<Codec>(estimatedFeeMethodName, messageFeeType.toHex());
-
-      // @ts-ignore
-      const estimatedFeeType = sourceApi.registry.createType('Option<Balance>', estimatedFeeCall);
-      const estimatedFee = estimatedFeeType.toString();
-
-      dispatchTransaction(TransactionActionCreators.setEstimateFee(estimatedFee));
-    };
-
-    if (areApiReady && payload) {
-      calculateFee();
-    }
-  }, [
-    areApiReady,
-    dispatchTransaction,
-    estimatedFeeMethodName,
-    laneId,
-    payload,
-    sourceApi.registry,
-    sourceApi.rpc.state,
-    targetChain
-  ]);
+    calculateFee(payload);
+  }, [calculateFee, payload]);
```

Comment:
Thanks! This make much more sense to me now and makes it easier to follow why and when things change.